### DISCUSSION TITLE
Feature Request: allow `CSV_BUFFER_SIZE` to be configurable

### DISCUSSION THREAD

**Discussion by pgguru** [duckdb/duckdb#14107] [Ideas]
(I debated characterizing this as a bug, but really it's a limitation I'm hoping to overcome.)

Hi, I am noticing some issues with CSV parsing when it comes to very large line sizes.  We have some data that exceeds the 32MB hard-coded limit defined in:

```
src/include/duckdb/execution/operator/csv_scanner/csv_buffer.hpp:78:    static constexpr idx_t CSV_BUFFER_SIZE = 32000000; // 32MB
```

This ends up throwing an internal error when trying to import a CSV file with longer lines.  While I recognize that this is not an issue for people using smaller CSV files, it seems like it should not be too difficult to make configurable, ideally via a SET GLOBAL option.  (I'm happy to leave the current value as the default setting here.)

This too-small buffer size also causes issues with the parallel reading, I assume due to basically having blocks without a complete line in them.  This end up hitting this code path in https://github.com/duckdb/duckdb/blob/v1.0.0/src/execution/operator/csv_scanner/scanner/string_value_scanner.cpp#L650

```
	// We need to check if we are getting the correct number of columns here.
	// If columns are correct, we add it, and that's it.
	if (cur_col_id != number_of_columns) {
		// We have too few columns:
		if (null_padding) {
```

and reporting about the wrong number of columns, despite the data in the line having the correct number of columns.  (I assume this basically only sees a single long field that takes up the entirety of the block starting before a block boundary, and this is why it's not seeing a full line.)

(It might seem that a natural blocksize would be related to the `maximum_line_size`, or at least a function of it, `Max(32000000,maximum_line_size)` probably rounded up to the nearest megabyte or something.)

**Comment by pdet** [comment]
Ha, I confess I thought no one would have a CSV file with rows > 32mb. 😅

So, the buffer_size is already configurable through the `buffer_size` option.

e.g.,
```FROM read_csv('r.csv',  buffer_size=30)```

The reason it is not documented is that it is not intended for general use; it's mostly used for testing over small buffers. It has not been developed for your use case, so it might have performance impacts, and you might encounter parallelism issues.

**Reply by pgguru** [reply]
Hi, thanks for letting me know. That will at least give us a knob to verify that this fixes the current issue (though using an unsupported API is not something I prefer to count on).

Is this undocumented but stable, or might it end up becoming documented/stable soon?

**Reply by pdet** [reply]
I don't think the API will change, but the option has not been battle-tested for buffer sizes larger than the original intention, which was for testing purposes (specially of values in-between buffers).

Considering there are actual use cases outside of testing, it's probably something we should document and test beyond the original scope.

**Reply by pgguru** [reply]
I'm not sure this is working as expected; here is a simple test case that presumably should be handled by the `buffer_size` option, but still fails:

```
D copy (select 1 as one, repeat('a', 40000000) r from generate_series(1,50)) to '/tmp/bigval.csv';
100% ▕████████████████████████████████████████████████████████████▏
D copy (select * from read_csv('/tmp/bigval.csv', auto_detect=False, parallel=False, columns={'id':'int', 'bigval':'text'}, delim=',', quote='"', escape='"', header=True, maximum_line_size=40000003, buffer_size=60000000)) to '/tmp/bigval.parquet';
Error: Invalid Input Error: CSV Error on Line: 5
Expected Number of Columns: 2 Found: 3
  file=
  delimiter = , (Set By User)
  quote = " (Set By User)
  escape = " (Set By User)
  new_line = \n (Auto-Detected)
  header = true (Set By User)
  skip_rows = 0 (Auto-Detected)
  date_format =  (Auto-Detected)
  timestamp_format =  (Auto-Detected)
  null_padding=0
  sample_size=20480
  ignore_errors=0
  all_varchar=0
```

Can you tell if I'm doing something wrong, or would you expect this to work as demonstrated?

**Reply by pdet** [reply]
Ah, my bad, I've actually implemented the `buffer_size` option to only be considered if the value is [lower than 8Mb](https://github.com/duckdb/duckdb/blob/main/src/execution/operator/csv_scanner/buffer_manager/csv_buffer_manager.cpp#L19), since that's what I cared about for testing.

I'll have a look in the future at implementing proper support for allowing it to be an arbitrary value, and to be user ready.

So currently that is really no workaround on reading a CSV file where the largest line doesn't fit in one buffer.

**Reply by pgguru** [reply]
Thanks for the update there; definitely having that selectable would be useful in the future.  I think the extra fields thing is a separate issue—do you want a new issue for that or can I continue to piggy-back off of #14111? (I don't know that I have the same ask as that user, but the underlying issue appears to be the same: there is a bug in the field detection.)

**Reply by pdet** [reply]
Unfortunately, I think his issue is entirely different.

**Comment by pgguru** [comment]
It looks like the latest release has fixed the underlying issue with the longer CSV lines, so closing this discussion.
